### PR TITLE
Add support for global prep commands

### PR DESCRIPTION
**Pull request by pgrunzjr**
## Description
Add the ability to add a global set of Prep Commands that will run for _all_ applications. Global prep configuration is accessible in the UI through new tab on Applications page. Users can also enabled/disabled execution of global commands on a per applications basis. Default is Enabled.

### Screenshot
Global Command Preparations on Configuration -> General page
image

Per-App Toggle of global prep command execution
image

### Issues Fixed or Closed
NA

## Type of Change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Dependency update (updates to dependencies)
- [x] Documentation update (changes to documentation)
- [ ] Repository update (changes to repository files, e.g. `.github/...`)

## Checklist
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added or updated the in code docstring/documentation-blocks for new or existing methods/components

## Branch Updates
LizardByte requires that branches be up-to-date before merging. This means that after any PR is merged, this branch
must be updated before it can be merged. You must also
[Allow edits from maintainers](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [ ] I want maintainers to keep my branch updated

### CODE REVIEW

File: src/process.cpp [hunk 1]
```diff
@@ -128,27 +123,61 @@ int proc_t::execute(int app_id) {
     terminate();
   });
 
-  for(; _undo_it != std::end(proc.prep_cmds); ++_undo_it) {
-    auto &cmd = _undo_it->do_cmd;
+  // Execute global prep commands if enabled
+  if (_app.include_global_prep)
```
Comment (Nonary) [inline src/process.cpp:127]: What was the reasoning for an additional for loop in this case? Why not just create a new iterator by merging them together? You can decide to just not add it to the new iterator if the checkmark is false.

File: src/process.cpp [hunk 2]
```diff
@@ -49,14 +49,6 @@ void process_end(bp::child &proc, bp::group &proc_handle) {
   proc.wait();
 }
 
-int exe_with_full_privs(const std::string &cmd, bp::environment &env, file_t &file, std::error_code &ec) {
```
Comment (Nonary) [inline src/process.cpp:52]: This is a great idea, but honestly this should be an entirely different pull request. Since it is more of a security concern in general, it's not ideal to mix multiple features or changes in one pull request generally.

File: src/process.cpp [hunk 3]
```diff
@@ -203,25 +232,59 @@ void proc_t::terminate() {
...
+    
+    child.wait();    
+    auto ret = child.exit_code();
 
     if(ret != 0) {
       BOOST_LOG(warning) << "Return code ["sv << ret << ']';
     }
   }
+  
+  // Execute global prep commands if enabled
```
Comment (Nonary) [inline src/process.cpp:260]: Much like the previous comment, we probably don't need an extra loop here.

File: src/process.cpp [hunk 4]
```diff
@@ -439,28 +502,49 @@ std::optional<proc::proc_t> parse(const std::string &file_name) {
...
       proc::ctx_t ctx;
 
-      auto prep_nodes_opt     = app_node.get_child_optional("prep-cmd"s);
-      auto detached_nodes_opt = app_node.get_child_optional("detached"s);
-      auto output             = app_node.get_optional<std::string>("output"s);
-      auto name               = parse_env_val(this_env, app_node.get<std::string>("name"s));
-      auto cmd                = app_node.get_optional<std::string>("cmd"s);
-      auto image_path         = app_node.get_optional<std::string>("image-path"s);
-      auto working_dir        = app_node.get_optional<std::string>("working-dir"s);
+      auto prep_nodes_opt      = app_node.get_child_optional("prep-cmd"s);
```
Comment (Nonary) [inline src/process.cpp:540]: You know now looking over this, we can probably just refactor this section entirely instead of spamming null checks for every variable. Nothing wrong with what you did here though, just brainstorming some ideas.

File: src/process.h [hunk 5]
```diff
@@ -39,6 +39,8 @@ struct cmd_t {
  *    filename -- The output of the commands are appended to filename
  */
 struct ctx_t {
+  ctx_t() : include_global_prep(true) {}
```
Comment (Nonary) [inline src/process.h:42]: Preferably booleans should initialize at false, it's not a hard requirement but that is general convention. I would invert the boolean here to state disable_global_prep instead. That way you don't have to add an initializer and it matches with most of the booleans in this code base.

**Review by Nonary — COMMENTED** [review]
I have some questions about these changes, I am not an expert at C++ but just curious on the changes as opposed to other common programing structures.

File: src/process.h (hunk 5 above)
Comment (ReenigneArcher) [reply src/process.h:42]: I think they just didn't get around to updating it yet... We had the same discussion here: https://discord.com/channels/804382334370578482/1074377411069747250/1077688007659896832

File: src/process.cpp (hunk 1 above)
Comment (pgrunzjr) [reply src/process.cpp:127]: I actually had them combined previously, before adding the ability to exclude them per app, but split them to make things easier to debug when adding that feature in. There's really no reason they couldn't be combined though, so I may change it back now that you've pointed it out.

File: src/process.cpp (hunk 2 above)
Comment (pgrunzjr) [reply src/process.cpp:52]: Yeah, in hindsight, it probably should have been a separate PR. I'll let @ReenigneArcher make the call on it. I'm happy to split the execution changes into a separate PR if you want.

File: src/process.cpp (hunk 4 above)
Comment (pgrunzjr) [reply src/process.cpp:540]: Agreed. It does look like #584 does some significant refactoring here.

File: src/process.h (hunk 5 above)
Comment (pgrunzjr) [reply src/process.h]: Fixed in latest commit.

File: src/process.cpp (hunk 1 above)
Comment (Nonary) [reply src/process.cpp]: I was just thinking when someone adds the capability to reference moonlight client variables (like client resolution, framerate, hdr setting, etc) it would be easier to implement if it's using an iterator. Multiple loops could work as well if you refactor it to use a function to do most of the work too

File: src/process.cpp (hunk 2 above)
Comment (Nonary) [reply src/process.cpp:52]: Well, as long as he agrees with it, #998 is the issue you can link for that PR

File: src/process.cpp (hunk 1 above)
Comment (pgrunzjr) [reply src/process.cpp]: I've updated it to load everything into a single vector for each app, so only a single loop is needed now.

**Comment by Nonary** [comment]
The latest changes look really good to me, although granted I am not fluent in C++. Just from an object oriented perspective, it looks good to me. I think all we really need now is documentation updates and we can merge it into nightly from there.

File: src/process.cpp (hunk 2 above)
Comment (ReenigneArcher) [reply src/process.cpp:52]: I agree that this change should be a separate PR.

File: src/process.cpp (hunk 4 above)
Comment (ReenigneArcher) [reply src/process.cpp:464]: That PR still needs a lot of work, so I wouldn't base any changes on what's in that to be honest.

For one it completely removes the UI from Sunshine which is a no go.

File: src/confighttp.cpp [hunk 6]
```diff
@@ -488,6 +488,45 @@ void uploadCover(resp_https_t response, req_https_t request) {
   outputTree.put("path", path);
 }
 
+void savePrepCmd(resp_https_t response, req_https_t request) {
```
Comment (ReenigneArcher) [inline src/confighttp.cpp:491]: Unfortunately this is probably not an easy change, but... instead of adding a new endpoint and function... can we just use the saveConfig endpoint and attach them to the config object?

**Review by ReenigneArcher — CHANGES_REQUESTED** [review]

**Comment by Nonary** [comment]
@pgrunzjr The global prep commands tab isn't working with the latest commits. I click on it and nothing happens.

File: src/process.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (pgrunzjr) [reply src/process.cpp:52]: Created a separate PR and will remove these change form this one.

#1022

File: src/confighttp.cpp (hunk 6 above)
Comment (pgrunzjr) [reply src/confighttp.cpp:491]: Might be a bit of a challenge since sunshine.conf isn't JSON, but I should be able to make it work. Would we also want to move the global prep tab onto the config page then? I feel like that makes more sense anyway.

Comment (ReenigneArcher) [reply src/confighttp.cpp:491]: > Would we also want to move the global prep tab onto the config page then?

Yea, I think it makes the most sense.

There's a couple of options that are like a json list (resolution and fps I believe), just going from memory, so that might not be 100% accurate.

Comment (pgrunzjr) [reply src/confighttp.cpp]: Now stored in config and setup moved to the config page.
image

**Comment by pgrunzjr** [comment]
> @pgrunzjr The global prep commands tab isn't working with the latest commits. I click on it and nothing happens.

Newer changes have moved global prep commands to the Configuration section and the Apps page tabs removed.

File: src/config.cpp [hunk 7]
```diff
@@ -424,7 +425,8 @@ sunshine_t sunshine {
   platf::appdata().string() + "/sunshine.conf", // config file
   {},                                           // cmd args
   47989,
-  platf::appdata().string() + "/sunshine.log", // log file
+  platf::appdata().string() + "/sunshine.log", // log file,
```
Comment (ReenigneArcher) [inline src/config.cpp:428]: comma not necessary after comment

File: src/config.cpp [hunk 8]
```diff
@@ -424,7 +425,8 @@ sunshine_t sunshine {
...
+  {}
```
Comment (ReenigneArcher) [inline src/config.cpp:429]: please add comment

**Review by ReenigneArcher — CHANGES_REQUESTED** [review]
Requesting a few small changes. We're going to need to add boost/json to our build process as well. I will take care of that in another PR so you don't have to dig through all the CI and packaging files.

I like the idea of using that though, can probably move the `resolutions` and `fps` options to use it as well, instead of that manual serialize function. I have some tweaks to how the config is handled in the UI in #1023, but I'll merge yours first so you don't have to fix any conflicts.

Edit: Can you also update the documentation, specifically "advanced_usage.rst". This is where we document each setting.

Edit 2: Apparently there's no `libboost-json-dev` library for debian distros. Do we need to install something additional?

**Comment by pgrunzjr** [comment]
> Edit 3: boost-json was introduced in Boost 1.75. This is going to be a problem since Ubuntu 22.04 is on 1.74, and we're building Ubuntu 20.04 with Boost 1.71.

For the purpose of this PR, I think I should be able to update it to use Boost PropertyTree for the JSON stuff for now. In the future, we could look into using another library for JSON, maybe https://github.com/nlohmann/json.

**Comment by pgrunzjr** [comment]
PR updated with documentation updates and to use Boost PropertyTree instead of Boost JSON.

File: src_assets/common/assets/web/config.html
Comment (ReenigneArcher) [inline src_assets/common/assets/web/config.html:1]: I think we need one minor change due a change in #1023. Around L966, just add the default value for the global prep commands. That might make L1000 irrelevant... Anyway the idea with #1023 was if the setting in the UI matches the default then we won't write it to the config file.

Comment (pgrunzjr) [reply src_assets/common/assets/web/config.html:1]: Defaults have been added. L1000 still needs to exist, otherwise Vue doesn't have anything to bind too and the page doesn't show the new rows when you click the + button unless you manually refresh.

File: src/config.cpp [hunk 9]
```diff
@@ -7,6 +7,9 @@
 
 #include <boost/asio.hpp>
 #include <boost/filesystem.hpp>
+#include <boost/json.hpp>
```
[the remark this answers is not in the document]
Comment (pgrunzjr) [reply src/config.cpp:10]: Ah, missed removing that. Will fix momentarily.

**Review by ReenigneArcher — APPROVED** [review]
